### PR TITLE
Add parsing and prettier support for @example tags in liquid doc

### PR DESCRIPTION
**Pull request by EvilGenius13**
## What are you adding in this PR?

References: https://github.com/Shopify/develop-advanced-edits/issues/471

Added support for parsing and formatting `@example` tags within `LiquidDoc` blocks.

## Feedback
Right now, we aren't controlling whitespace for example tags. If someone wants to write `check   out.     my great    example`, it would stay in that format. Should trim whitespace or allow people writing examples to have full control?

## What's next? Any followup issues?

I want to add prettier support to automatically move `@example` tags and content below `@param` as it is the proper flow.

## What did you learn?

AST's and CST's are quite puzzle to figure out.

## Test
https://github.com/user-attachments/assets/70c8452b-820d-4091-acbb-dd6a720821cc

## Before you deploy

- [ ] I included a minor bump `changeset`
- [ ] My feature is backward compatible

### CODE REVIEW

File: packages/liquid-html-parser/src/stage-1-cst.spec.ts [hunk 1]
```diff
@@ -1108,6 +1108,83 @@ describe('Unit: Stage 1 (CST)', () => {
...
+        });
+
+        it('should parse example tag with content that has leading whitespace', () => {
+          const testStr = `{% doc %} @example         hello there       {%- enddoc %}`;
+          cst = toCST(testStr);
+          expectPath(cst, '0.type').to.equal('LiquidRawTag');
+          expectPath(cst, '0.name').to.equal('doc');
+          expectPath(cst, '0.children.0.type').to.equal('LiquidDocExampleNode');
+          expectPath(cst, '0.children.0.name').to.equal('example');
+          expectPath(cst, '0.children.0.exampleContent.value').to.equal('hello there');
```
Comment (jamesmengo) [inline packages/liquid-html-parser/src/stage-1-cst.spec.ts:1128]: let's add some tests for locStart / locEnd and document how we want that to behave, especially considering whitespace etc

Can also add some assertions in stage-2 for position

File: packages/prettier-plugin-liquid/src/printer/print/liquid.ts [hunk 2]
```diff
@@ -536,6 +537,30 @@ export function printLiquidDocParam(
...
+
+  if (node.exampleContent?.value) {
+    const content = node.exampleContent.value.trim();
+    if (content) {
+      parts.push(hardline);
+      const lines = content
+        .split('\n')
+        .map((line) => line.trim())
+        .filter(Boolean);
+      parts.push(join(hardline, lines));
```
Comment (jamesmengo) [inline packages/prettier-plugin-liquid/src/printer/print/liquid.ts:557]: might be able to clean this up depending on how we end up handling the trimming in earlier stages

**Review by jamesmengo — APPROVED** [review]

File: packages/prettier-plugin-liquid/src/printer/print/liquid.ts [hunk 3]
```diff
@@ -536,6 +537,30 @@ export function printLiquidDocParam(
...
+  const parts: Doc[] = ['@example'];
+
+  if (node.exampleContent?.value) {
+    const content = node.exampleContent.value.trim();
+    if (content) {
+      parts.push(hardline);
+      const lines = content
+        .split('\n')
+        .map((line) => line.trim())
+        .filter(Boolean);
```
Comment (jamesmengo) [inline packages/prettier-plugin-liquid/src/printer/print/liquid.ts:556]: this behaviour should be tested in the `test/liquid-doc` folder

File: packages/liquid-html-parser/src/stage-2-ast.ts [hunk 4]
```diff
@@ -765,6 +766,14 @@ export interface LiquidDocParamNode extends ASTNode<NodeTypes.LiquidDocParamNode
   /** Optional type annotation for the parameter (e.g. "{string}", "{number}") */
   paramType: TextNode | null;
 }
+
+/** Represents a `@example` node in a LiquidDoc comment - `@example @exampleContent` */
```
Comment (jamesmengo) [inline packages/liquid-html-parser/src/stage-2-ast.ts:770]:
```suggestion
/** Represents a `@example` node in a LiquidDoc comment - `@example exampleContent` */
```

File: packages/liquid-html-parser/src/stage-2-ast.ts [hunk 5]
```diff
@@ -1297,6 +1306,22 @@ function buildAst(
...
+          type: NodeTypes.LiquidDocExampleNode,
+          name: node.name,
+          position: position(node),
+          source: node.source,
+          exampleContent: {
+            type: NodeTypes.TextNode,
+            value: node.exampleContent.value,
+            position: position(node.exampleContent),
+            source: node.exampleContent.source,
+          },
```
Comment (jamesmengo) [inline packages/liquid-html-parser/src/stage-2-ast.ts:1320]:
```suggestion
          exampleContent: toTextNode(node.exampleContent),
```

Comment (jamesmengo) [reply packages/liquid-html-parser/src/stage-2-ast.ts:1320]: Could you also fix this for `paramNode`? A follow-up PR is fine too :)

File: packages/prettier-plugin-liquid/src/printer/print/liquid.ts [hunk 6]
```diff
@@ -536,6 +537,62 @@ export function printLiquidDocParam(
...
+  options: LiquidParserOptions,
+  _print: LiquidPrinter,
+  _args: LiquidPrinterArgs,
+): Doc {
+  const node = path.getValue();
+  const parts: Doc[] = ['@example'];
+
+  if (node.exampleContent?.value) {
+    const content = node.exampleContent.value;
+    if (content) {
```
Comment (jamesmengo) [inline packages/prettier-plugin-liquid/src/printer/print/liquid.ts:551]:
```suggestion
    const content = node.exampleContent?.value
    if (content) {
        ....
    }
```

File: packages/prettier-plugin-liquid/src/test/liquid-doc/index.liquid [hunk 7]
```diff
@@ -27,3 +27,46 @@ It should normalize the param description
...
+  @example
+
+  Here is my content and a newline
+
+{% enddoc %}
+
+It should respect example content with param and description
+{% doc %}
+@param paramName - param with description
+@example
```
Comment (jamesmengo) [inline packages/prettier-plugin-liquid/src/test/liquid-doc/index.liquid:69]: I think @example should have any empty newline if it's not the first node, wdyt?

Something like [this](https://github.com/Shopify/theme-tools/commit/f39ac31976abedc4909443df51277aeef3f6d824)?

**Comment by jamesmengo** [comment]
Thanks for the changes! The only blocking comment I have is around [this](https://app.graphite.dev/github/pr/Shopify/theme-tools/725/Add-parsing-and-prettier-support-for-example-tags-in-liquid-doc#comment-PRRC_kwDOKAj0yc5zpinv), let me know your thoughts!

File: packages/liquid-html-parser/src/stage-2-ast.ts (hunk 5 above)
Comment (EvilGenius13) [reply packages/liquid-html-parser/src/stage-2-ast.ts:1320]: Fixed in https://github.com/Shopify/theme-tools/pull/759
Left out `paramNode` as it's currently being refactored in another PR

File: packages/prettier-plugin-liquid/src/printer/print/liquid.ts (hunk 6 above)
Comment (EvilGenius13) [reply packages/prettier-plugin-liquid/src/printer/print/liquid.ts:551]: Fixed in https://github.com/Shopify/theme-tools/pull/759

File: packages/prettier-plugin-liquid/src/test/liquid-doc/index.liquid (hunk 7 above)
Comment (EvilGenius13) [reply packages/prettier-plugin-liquid/src/test/liquid-doc/index.liquid:69]: Agreed. This will be tackled in a separate PR as there is some other prettier logic that will need refactoring as well.
